### PR TITLE
Add metrics for log flush rate and relevant queue time and sizes.

### PR DESCRIPTION
**Pull request by efeg**
Add the following metrics:

`REQUEST_QUEUE_SIZE`
`RESPONSE_QUEUE_SIZE`
`PRODUCE_REQUEST_QUEUE_TIME_MS_MAX`
`PRODUCE_REQUEST_QUEUE_TIME_MS_MEAN`
`CONSUMER_FETCH_REQUEST_QUEUE_TIME_MS_MAX`
`CONSUMER_FETCH_REQUEST_QUEUE_TIME_MS_MEAN`
`FOLLOWER_FETCH_REQUEST_QUEUE_TIME_MS_MAX`
`FOLLOWER_FETCH_REQUEST_QUEUE_TIME_MS_MEAN`
`PRODUCE_TOTAL_TIME_MS_MAX`
`PRODUCE_TOTAL_TIME_MS_MEAN`
`CONSUMER_FETCH_TOTAL_TIME_MS_MAX`
`CONSUMER_FETCH_TOTAL_TIME_MS_MEAN`
`FOLLOWER_FETCH_TOTAL_TIME_MS_MAX`
`FOLLOWER_FETCH_TOTAL_TIME_MS_MEAN`
`LOG_FLUSH_RATE`
`LOG_FLUSH_TIME_MS_MAX`
`LOG_FLUSH_TIME_MS_MEAN`
`PRODUCE_LOCAL_TIME_MS_MAX`
`PRODUCE_LOCAL_TIME_MS_MEAN`
`CONSUMER_FETCH_LOCAL_TIME_MS_MAX`
`CONSUMER_FETCH_LOCAL_TIME_MS_MEAN`
`FOLLOWER_FETCH_LOCAL_TIME_MS_MAX`
`FOLLOWER_FETCH_LOCAL_TIME_MS_MEAN`

### CODE REVIEW

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/BrokerMetricSample.java [hunk 1]
```diff
@@ -166,10 +308,25 @@ public double allTopicsFetchRequestRate() {
    * 8 bytes - all topics produce request rate
    * 8 bytes - all topics fetch request rate
    * 8 bytes - sample time
+   * 8 bytes - broker request queue size
```
Comment (becketqin) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/BrokerMetricSample.java]: Probably does not matter much in our case as we won't have many brokers. But the queue size and time can be int. Unlike request rate/bytes in rate, they are unlikely to exceed the integer range.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/BrokerMetricSample.java [hunk 2]
```diff
@@ -166,10 +308,25 @@ public double allTopicsFetchRequestRate() {
...
+   * 8 bytes - broker response queue size
+   * 8 bytes - broker produce request queue size ms (max)
```
Comment (becketqin) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/BrokerMetricSample.java]: queue time ms

File: cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/metric/MetricsUtils.java [hunk 3]
```diff
@@ -150,19 +176,13 @@ public static boolean isInterested(com.yammer.metrics.core.MetricName metricName
...
-        && INTERESTED_REQUEST_TYPE.contains(tags.get(REQUEST_TYPE_KEY))) {
-      return true;
-    } else if (group.equals(KAFKA_LOG)
-        && LOG_GROUP.equals(type)
-        && INTERESTED_LOG_METRIC_NAMES.contains(name)) {
-      return true;
+      return (INTERESTED_TOPIC_METRIC_NAMES.contains(name) && BROKER_TOPIC_METRICS_GROUP.equals(type)) || (
+          INTERESTED_SERVER_METRIC_NAMES.contains(name) && REQUEST_KAFKA_HANDLER_POOL_GROUP.equals(type));
+    } else if (group.equals(KAFKA_NETWORK) && INTERESTED_NETWORK_METRIC_NAMES.contains(name)) {
+      return REQUEST_CHANNEL_GROUP.equals(type)
```
[the remark this answers is not in the document]
Comment (efeg) [reply cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/metric/MetricsUtils.java]: `INTERESTED_NETWORK_METRIC_NAMES` specifies the `RequestQueueSize`  or `ResponseQueueSize`.

File: cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/metric/MetricType.java [hunk 4]
```diff
@@ -34,7 +34,24 @@
   BROKER_PRODUCE_REQUEST_RATE((byte) 16),
   BROKER_CONSUMER_FETCH_REQUEST_RATE((byte) 17),
   BROKER_FOLLOWER_FETCH_REQUEST_RATE((byte) 18),
-  BROKER_REQUEST_HANDLER_AVG_IDLE_PERCENT((byte) 19);
+  BROKER_REQUEST_HANDLER_AVG_IDLE_PERCENT((byte) 19),
+  BROKER_REQUEST_QUEUE_SIZE((byte) 20),
+  BROKER_RESPONSE_QUEUE_SIZE((byte) 21),
+  BROKER_PRODUCE_REQUEST_QUEUE_TIME_MS_MAX((byte) 22),
+  BROKER_PRODUCE_REQUEST_QUEUE_TIME_MS_MEAN((byte) 23),
+  BROKER_CONSUMER_FETCH_REQUEST_QUEUE_TIME_MS_MAX((byte) 24),
```
Comment (becketqin) [inline cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/metric/MetricType.java]: We might want to add the local time as well.

File: cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/metric/MetricType.java [hunk 5]
```diff
@@ -34,8 +34,30 @@
...
+  BROKER_PRODUCE_TOTAL_TIME_MS_MAX((byte) 28),
+  BROKER_PRODUCE_TOTAL_TIME_MS_MEAN((byte) 29),
+  BROKER_CONSUMER_FETCH_TOTAL_TIME_MS_MAX((byte) 30),
+  BROKER_CONSUMER_FETCH_TOTAL_TIME_MS_MEAN((byte) 31),
+  BROKER_FOLLOWER_FETCH_TOTAL_TIME_MS_MAX((byte) 32),
+  BROKER_FOLLOWER_FETCH_TOTAL_TIME_MS_MEAN((byte) 33),
+  BROKER_LOG_FLUSH_RATE((byte) 34),
+  BROKER_LOG_FLUSH_TIME_MS_MAX((byte) 35),
+  BROKER_LOG_FLUSH_TIME_MS_MEAN((byte) 36),
+  BROKER_PRODUCE_LOCAL_TIME_MS_MAX((byte) 37),
```
Comment (becketqin) [inline cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/metric/MetricType.java]: Nit: Can we move the log flush rate/time to be after the local time so the request related times sits together.

**Pull request merged**